### PR TITLE
Add `-Wperformance` lints for `*Ref` boxing and nonlocal `return`

### PR DESCRIPTION
**Pull request by som-snytt**
Warn when local variables are captured by lambdas or by-name args and moved to the heap.

The warnings are behind a new `-Wperformance` flag, as the user did nothing wrong but might be surprised by a difference in performance.

Similarly, `-Xlint:nonlocal-return` is also enabled by the new flag, though it is not removed from `-Xlint`.

Revives https://github.com/scala/scala/pull/8691 by @hrhino.

### CODE REVIEW

**Comment by som-snytt** [comment]
It would be nice to report whether boxings are eliminated by optimizer after closure inlining.

**Comment by SethTisue** [comment]
@som-snytt I don't doubt this will be merged in the end, but would you mind filling in the PR description, to aid reviewers? (But also as end-user documentation, since this is release-notable.)

**Review by SethTisue — APPROVED** [review]

**Comment by SethTisue** [comment]
Let's wait to merge until @lrytz is available.

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 1]
```diff
@@ -505,6 +505,9 @@ abstract class LambdaLift extends InfoTransform {
               }
             }
 
+            if (settings.warnCaptured)
+              reporter.warning(tree.pos, s"Modification of variable $name within a closure causes it to be boxed.")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala]: Could you use `runReporting.warning` to not bypass `Wconf`?

**Comment by som-snytt** [comment]
This is under-specified, but 2.13.8 says
```
scala> StringContext.glob(Nil,"")
java.lang.NegativeArraySizeException: -1
  at scala.StringContext$.glob(StringContext.scala:233)
  ... 32 elided
```
and 2.13.9 will say
```
scala> StringContext.glob(Nil,"")
java.util.NoSuchElementException: head of empty list
  at scala.collection.immutable.Nil$.head(List.scala:662)
  at scala.collection.immutable.Nil$.head(List.scala:661)
  at scala.StringContext$.glob(StringContext.scala:229)
  ... 34 elided
```

**Comment by som-snytt** [comment]
Those other commits must have been prompted by the warning. But IIRC they are a balance of style rather than benchmark-driven.

Somewhere I must have seen someone's suggestion for the hand-rolled function object.

**Comment by SethTisue** [comment]
I see that back in February 2020 my take on Harrison's PR was:

> agree this more properly would be part of a separate linter or Scalafix ruleset

and, well, I still wonder about that a bit. but I don't recall my whole thought process, there.

the implementation doesn't seem like a likely maintenance burden. and perhaps my previous thinking was merely pedantic.

File: src/library/scala/collection/Iterable.scala [hunk 2]
```diff
@@ -597,11 +597,13 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
       val bldr = m.getOrElseUpdate(k, iterableFactory.newBuilder[B])
       bldr += f(elem)
     }
-    var result = immutable.Map.empty[K, CC[B]]
-    m.foreach { case (k, v) =>
-      result = result + ((k, v.result()))
+    object result extends Function[(K, Builder[B, CC[B]]), Unit] {
```
Comment (xuwei-k) [inline src/library/scala/collection/Iterable.scala:600]: use `class` and `AbstractFunction` ? 🤔 

- https://github.com/scala/scala/pull/9216

Comment (som-snytt) [reply src/library/scala/collection/Iterable.scala:600]: Thanks, I will also add a lint for that.

**Pull request merged**